File: benches/benchmark.cpp
#include <filesystem>
#include <fstream>

#include <catch2/benchmark/catch_benchmark.hpp>
#include <catch2/catch_test_macros.hpp>

#include "../sources/algorithm.hpp"

TEST_CASE("Benchmark", "[bench]")
{
    spdlog::set_level(spdlog::level::off);
    for (const auto& entry : std::filesystem::directory_iterator("data/br_json"))
    {
        if (entry.path().string().find("001") == std::string::npos)
        {
            continue;
        }
        BENCHMARK(entry.path().filename().string())
        {
            Input input = json::parse(std::ifstream(entry.path()));
            Output output = Algorithm(input).run();
            return output.containers.size(); // 返回方案数量，防止优化
        };
    }
}

File: benches/benchmark.rs
use criterion::{Criterion, criterion_group, criterion_main};
use pack3d::algorithm::Algorithm;
use pack3d::tool::parse_input;
use std::fs;

fn benchmark_algorithm(c: &mut Criterion) {
    for entry in fs::read_dir("data/br_json").unwrap() {
        let path = entry.unwrap().path();

        if !path.to_str().unwrap().contains("001") {
            continue;
        }

        let filename = path.file_name().unwrap().to_str().unwrap().to_string();
        let path_str = path.to_str().unwrap().to_string();

        c.bench_function(&filename, |b| {
            b.iter(|| {
                let input = parse_input(&path_str);
                let output = Algorithm::new(input).run();
                std::hint::black_box(output.containers.len())
            });
        });
    }
}

criterion_group!(benches, benchmark_algorithm);
criterion_main!(benches);
